src/routers/alpha-router/alpha-router.ts

```diff
@@ -119,51 +119,91 @@ export class AlphaRouter implements IRouter<AlphaRouterConfig>, ISwapToRatio<Alp
-        sqrtPriceX96,
-        sqrtPriceX96Lower,
-        precision,
-        true
+      let optimalRatio = this.calculateOptimalRatio(
+        position,
+        position.pool.sqrtRatioX96
       )
 
-      const amountToSwap = calculateRatioAmountIn(
-        new Fraction(token0Proportion, token1Proportion),
-        position.pool.token0Price,
-        token0Balance,
-        token1Balance,
-      )
+      let zeroForOne: boolean
+      if (position.pool.tickCurrent < position.tickLower) {
+        zeroForOne = true
+      } else if (position.pool.tickCurrent > position.tickUpper) {
+        zeroForOne = false
+      } else {
+        zeroForOne = new Fraction(token0Balance.quotient, token1Balance.quotient).greaterThan(optimalRatio)
+        if (!zeroForOne) optimalRatio = optimalRatio.invert()
+      }
 
-      return this.routeExactIn(
-        amountToSwap.currency,
-        amountToSwap.currency == token0Balance.currency
-          ? token1Balance.currency
-          : token0Balance.currency,
-        amountToSwap,
-        swapConfig,
-        routingConfig
-      )
+      const [inputBalance, outputBalance] = zeroForOne
+        ? [token0Balance, token1Balance]
+        : [token1Balance, token0Balance]
+
+      let exchangeRate: Fraction = zeroForOne
+        ? position.pool.token0Price
+        : position.pool.token1Price
+      let swap: SwapRoute<TradeType.EXACT_INPUT> | null = null
+      let ratioAchieved = false
+      let n = 0
+
+      while (!ratioAchieved) {
+        n++
+        if (n > maxIterations) {
+          return null;
+        }
+
+        let amountToSwap = calculateRatioAmountIn(
+          optimalRatio,
+          exchangeRate,
+          inputBalance,
+          outputBalance,
+        )
+
+        swap = await this.routeExactIn(
+          inputBalance.currency,
+          outputBalance.currency,
+          amountToSwap,
+          swapConfig,
+          routingConfig
+        )
+        if (!swap) {
+          return null;
+        }
+
+        let inputBalanceUpdated = inputBalance.subtract(swap.trade.inputAmount)
+        let outputBalanceUpdated = outputBalance.add(swap.trade.outputAmount)
+        let newRatio = inputBalanceUpdated.divide(outputBalanceUpdated)
+
+        swap.route.forEach(route => {
+          route.route.pools.forEach((pool, i) => {
+            if(
+              pool.token0.equals(position.pool.token0) &&
+              pool.token1.equals(position.pool.token1) &&
+              pool.fee == position.pool.fee
+            ) {
+              optimalRatio = this.calculateOptimalRatio(
+                position,
+                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
+              )
+            }
+          })
+        })
+
+        ratioAchieved = this.absoluteValue(
+          newRatio.asFraction.divide(optimalRatio).subtract(1)
+        ).lessThan(errorTolerance)
+        exchangeRate = swap.trade.outputAmount.divide(swap.trade.inputAmount)
+      }
```

Comment:
Would be great to log our newRatio, optimalRatio, and iteration number here

Reply:
added some logs, lmk if I should add anything, change the format!